letsencrypt/client/tests/standalone_authenticator_test.py

```diff
@@ -0,0 +1,564 @@
+        result = self.authenticator.perform([chall1, chall2, bad_chall])
+        self.assertEqual(len(self.authenticator.tasks), 2)
+        self.assertTrue(
+            self.authenticator.tasks.has_key("foononce.acme.invalid"))
+        self.assertTrue(
+            self.authenticator.tasks.has_key("barnonce.acme.invalid"))
+        self.assertTrue(isinstance(result, list))
+        self.assertEqual(len(result), 3)
+        self.assertEqual(result, [None, None, False])
+        self.authenticator.start_listener.assert_called_once_with(443, key)
+
+    def test_perform_with_pending_tasks(self):
+        self.authenticator.tasks = {"foononce.acme.invalid": "cert_data"}
+        extra_challenge = DvsniChall("a", "b", "c", "d")
+        self.assertRaises(Exception, self.authenticator.perform,
+                          [extra_challenge])
+
+    def test_perform_without_challenge_list(self):
+        extra_challenge = DvsniChall("a", "b", "c", "d")
+        # This is wrong because a challenge must be specified.
+        self.assertRaises(Exception, self.authenticator.perform, [])
+        # This is wrong because it must be a list, not a bare challenge.
+        self.assertRaises(Exception, self.authenticator.perform,
+                          extra_challenge)
+        # This is wrong because the list must contain at least one challenge.
+        self.assertRaises(Exception, self.authenticator.perform,
+                          range(20))
+
+
+class StartListenerTest(unittest.TestCase):
+    """Tests for start_listener() method."""
+    def setUp(self):
+        from letsencrypt.client.standalone_authenticator import \
+            StandaloneAuthenticator
+        self.authenticator = StandaloneAuthenticator()
+
+    @mock.patch("letsencrypt.client.standalone_authenticator."
+                "Crypto.Random.atfork")
+    @mock.patch("letsencrypt.client.standalone_authenticator.os.fork")
+    def test_start_listener_fork_parent(self, mock_fork, mock_atfork):
+        self.authenticator.do_parent_process = mock.Mock()
+        mock_fork.return_value = 22222
+        self.authenticator.start_listener(1717, "key")
+        self.assertEqual(self.authenticator.child_pid, 22222)
+        self.authenticator.do_parent_process.assert_called_once_with(1717)
+        mock_atfork.assert_called_once_with()
+
+    @mock.patch("letsencrypt.client.standalone_authenticator."
+                "Crypto.Random.atfork")
+    @mock.patch("letsencrypt.client.standalone_authenticator.os.fork")
+    def test_start_listener_fork_child(self, mock_fork, mock_atfork):
+        import os
+        self.authenticator.do_parent_process = mock.Mock()
+        self.authenticator.do_child_process = mock.Mock()
+        mock_fork.return_value = 0
+        self.authenticator.start_listener(1717, "key")
+        self.assertEqual(self.authenticator.child_pid, os.getpid())
+        self.authenticator.do_child_process.assert_called_once_with(1717,
```

Comment:
I had hanging indents for *all of these* before, and I got rid of them because they were rejected by pep8. But @jdkasten says that we can use them because pylint accepts them even if pep8 doesn't.